From 1cdd4c7d79a832421c7b1225425446727ab20acf Mon Sep 17 00:00:00 2001
From: Lubomir Rintel <lkundrak@v3.sk>
Date: Mon, 17 Jun 2019 13:41:05 +0200
Subject: [PATCH 1/4] build/autotools: generate "config-extra.h" via makefile
 "config-extra.h.mk"

When the code that generates "config-extra.h" changes, we want to regenerate
the file. Move that code to a separate makefile so we can add a
dependency.

Otherwise, we'd had to depend on "Makefile", which itself is generated by
Makefile.am.

Also, depend on "config.h" to regenerate it when ./configure runs and
touches that header. This may not cover all cases where ./configure's
configuration changes and a regeneration would be due. But such is life.

Also, most components depend on this header, so let various .dirstamp
files depend on it, so we are sure to build this first. That because,
autotools generates dependencies for header files automatically, but
that requires that the header file exist. Such automatic dependencies
don't work out-of-the-box for generated headers.

Co-authored-by: Thomas Haller <thaller@redhat.com>

Upstream-status: Backport from
[https://cgit.freedesktop.org/NetworkManager/NetworkManager/commit/?id=11d59de600e4af093db642279945d79efa5d1f33]

Signed-off-by: Liu Haitao <haitao.liu@windriver.com>
---
 Makefile.am       | 40 +++++++++++-----------------------------
 config-extra.h.mk | 28 ++++++++++++++++++++++++++++
 2 files changed, 39 insertions(+), 29 deletions(-)
 create mode 100644 config-extra.h.mk

diff --git a/Makefile.am b/Makefile.am
index f594dd3..ba5be8e 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -109,38 +109,20 @@ dflt_cppflags = -std=gnu11
 
 ###############################################################################
 
-# with autotools, it is not easily possible to generate these defines
-# from configure.ac.
-#
-# See https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Installation-Directory-Variables.html
-#
-# With meson-only, they could just be set via config.h.meson
-config-extra.h: Makefile
-	$(AM_V_GEN) \
-	echo "/* Generated by Makefile.am */"                    >$@ && \
-	echo "#define BINDIR            \"$(bindir)\""          >>$@ && \
-	echo "#define DATADIR           \"$(datadir)\""         >>$@ && \
-	echo "#define LIBEXECDIR        \"$(libexecdir)\""      >>$@ && \
-	echo "#define LOCALSTATEDIR     \"$(localstatedir)\""   >>$@ && \
-	echo "#define NMCONFDIR         \"$(nmconfdir)\""       >>$@ && \
-	echo "#define NMLIBDIR          \"$(nmlibdir)\""        >>$@ && \
-	echo "#define NMLOCALEDIR       \"$(nmlocaledir)\""     >>$@ && \
-	echo "#define NMPLUGINDIR       \"$(plugindir)\""       >>$@ && \
-	echo "#define NMRUNDIR          \"$(nmrundir)\""        >>$@ && \
-	echo "#define NMSTATEDIR        \"$(nmstatedir)\""      >>$@ && \
-	echo "#define NMVPNDIR          \"$(pkglibdir)\""       >>$@ && \
-	echo "#define NM_BUILD_BUILDDIR \"$(abs_builddir)\""    >>$@ && \
-	echo "#define NM_BUILD_SRCDIR   \"$(abs_srcdir)\""      >>$@ && \
-	echo "#define PPPD_PLUGIN_DIR   \"$(PPPD_PLUGIN_DIR)\"" >>$@ && \
-	echo "#define PREFIX            \"$(prefix)\""          >>$@ && \
-	echo "#define RUNDIR            \"$(rundir)\""          >>$@ && \
-	echo "#define RUNSTATEDIR       \"$(runstatedir)\""     >>$@ && \
-	echo "#define SYSCONFDIR        \"$(sysconfdir)\""      >>$@ && \
-	true
+include config-extra.h.mk
 
 DISTCLEANFILES += config-extra.h
 
-libnm-core/.dirstamp: config-extra.h
+$(libnm_core_lib_h_pub_mkenums):    config-extra.h
+libnm-core/.dirstamp:               config-extra.h
+shared/.dirstamp:                   config-extra.h
+shared/nm-glib-aux/.dirstamp:       config-extra.h
+shared/nm-std-aux/.dirstamp:        config-extra.h
+shared/nm-udev-aux/.dirstamp:       config-extra.h
+shared/systemd/.dirstamp:           config-extra.h
+shared/systemd/src/basic/.dirstamp: config-extra.h
+src/dhcp/.dirstamp:                 config-extra.h
+ 
 
 ###############################################################################
 
diff --git a/config-extra.h.mk b/config-extra.h.mk
new file mode 100644
index 0000000..036de16
--- /dev/null
+++ b/config-extra.h.mk
@@ -0,0 +1,28 @@
+# With autotools, it is not easily possible to generate these defines
+# from configure.ac.
+#
+# See https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Installation-Directory-Variables.html
+#
+# With meson-only, they could just be set via config.h.meson
+config-extra.h: config-extra.h.mk config.h
+	$(AM_V_GEN) \
+	echo "/* Generated by Makefile.am */"                    >$@ && \
+	echo "#define BINDIR            \"$(bindir)\""          >>$@ && \
+	echo "#define DATADIR           \"$(datadir)\""         >>$@ && \
+	echo "#define LIBEXECDIR        \"$(libexecdir)\""      >>$@ && \
+	echo "#define LOCALSTATEDIR     \"$(localstatedir)\""   >>$@ && \
+	echo "#define NMCONFDIR         \"$(nmconfdir)\""       >>$@ && \
+	echo "#define NMLIBDIR          \"$(nmlibdir)\""        >>$@ && \
+	echo "#define NMLOCALEDIR       \"$(nmlocaledir)\""     >>$@ && \
+	echo "#define NMPLUGINDIR       \"$(plugindir)\""       >>$@ && \
+	echo "#define NMRUNDIR          \"$(nmrundir)\""        >>$@ && \
+	echo "#define NMSTATEDIR        \"$(nmstatedir)\""      >>$@ && \
+	echo "#define NMVPNDIR          \"$(pkglibdir)\""       >>$@ && \
+	echo "#define NM_BUILD_BUILDDIR \"$(abs_builddir)\""    >>$@ && \
+	echo "#define NM_BUILD_SRCDIR   \"$(abs_srcdir)\""      >>$@ && \
+	echo "#define PPPD_PLUGIN_DIR   \"$(PPPD_PLUGIN_DIR)\"" >>$@ && \
+	echo "#define PREFIX            \"$(prefix)\""          >>$@ && \
+	echo "#define RUNDIR            \"$(rundir)\""          >>$@ && \
+	echo "#define RUNSTATEDIR       \"$(runstatedir)\""     >>$@ && \
+	echo "#define SYSCONFDIR        \"$(sysconfdir)\""      >>$@ && \
+	true
-- 
2.23.0

